Bluetooth: controller: Guard against race in connection establishment

mtpr-ot opened the pull request:
In the time between a NODE_RX_TYPE_CONNECTION node is sent from LLL and
demuxed in ULL, an ADV role disable may be executed.
This makes the LLL data referenced in the node NULL/invald, and
ull_conn_setup would operate on invalid data.

This commit introduces a check in ADV disable to disallow the operation
(including conn invalidation), if a connection has been initiated.

To prevent pipeline-queued prepares from advertising after disable has
been initiated, set 'cancelled' flag for immediate signalling to LLL.

Fixes: #33868

Signed-off-by: Morten Priess <mtpr@oticon.com>

mtpr-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -807,32 +807,43 @@ bool ull_conn_peer_connected(uint8_t own_addr_type, uint8_t *own_addr,
-	switch (lll->role) {
+	if (lll) {
+		conn = lll->hdr.parent;
+		if ((lll == &conn->lll) && (lll->handle == 0xffff)) {
```
This is a bit "over the top" and not necessary for this particualr race, but always nice to be safe...

cvinayak reviewed (changes requested):
Connection *established* shall be prioritized over advertisement disable, the fix has to be in the `disable()` to detect the generated `NODE_RX_TYPE_CONNECTION` and fail with the HCI command response as cmd disallowed.

mtpr-ot commented:
> Connection _established_ shall be prioritized over advertisement disable, the fix has to be in the `disable()` to detect the generated `NODE_RX_TYPE_CONNECTION` and fail with the HCI command response as cmd disallowed.

OK, good point. Will try to rework.

mtpr-ot commented:
@cvinayak, I'm now using the `slave.initiated` flag in `lll_adv->conn` to detect connection initiation in `disable` - like in `prepare_cb`. Do you agree that would be safe?

cvinayak commented:
> 
> 
> @cvinayak, I'm now using the `slave.initiated` flag in `lll_adv->conn` to detect connection initiation in `disable` - like in `prepare_cb`. Do you agree that would be safe?

I feel like talking over the race conditions of ISR/LLL that sets up the expected reception of the CONNECT_IND in LLL that could happen before the `is_enabled` flags is checked, that could happen after the `is_enabled` check.  Do you think a handshake is not necessary, like being done when central is cancelling connection?

mtpr-ot committed:
Bluetooth: controller: Guard against race in conn. establishment

In the time between a NODE_RX_TYPE_CONNECTION node is sent from LLL and
demuxed in ULL, an ADV role disable may be executed.
This makes the LLL data referenced in the node NULL/invald, and
ull_conn_setup would operate on invalid data.

This commit introduces a check in ADV disable to disallow the operation
(including conn invalidation), if a connection has been initiated.

To prevent pipeline-queued prepares from advertising after disable has
been initiated, set 'cancelled' flag for immediate signalling to LLL.

Signed-off-by: Morten Priess <mtpr@oticon.com>
Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Remove unused code in auxiliary advertising

Remove unused early prototyping code experimenting with
piggyback of auxiliary PDUs.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak reviewed (approved):
I have forced push my suggestions, please pull my changes, review and verify at your end.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:
```diff
@@ -741,10 +741,11 @@ static int prepare_cb(struct lll_prepare_param *p)
 	 */
-	if (unlikely(lll->conn && lll->conn->slave.initiated)) {
+	if (unlikely(lll->conn &&
+		(lll->conn->slave.initiated || lll->conn->slave.cancelled))) {
```
Have updated myself and force push to this PR.

mtpr-ot commented:
Additions look good, and local tests are passing.

kruithofa reviewed (approved):
LGTM

mtpr-ot commented:
@cvinayak maybe we should trigger a CI rerun. Seems to have been fixed.
